File: Javascript/Map/3_Longest_Substring_Without_Repeating_Characters.js
/*
    Given a string s
    Find the length of the longest substring without duplicate characters.
*/

/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
    let maxLen = 0;
    let left = 0;
    let right = 0;
    const found = new Map();

    while (right < s.length) {
        // new character or last occurrence is before left border 
        if (!found.has(s[right]) || found.get(s[right]) < left) {
            found.set(s[right], right);
            maxLen = Math.max(right - left + 1, maxLen);
        } else {
            // update left to be after the last occurrence of the current character
            left = found.get(s[right]) + 1;
            found.set(s[right], right);
        }
        right++;
    }

    return maxLen;
};

File: Problem-lists/Algorithm_Patterns/Sliding_Window/3_Longest_Substring_Without_Repeating_Characters.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;

/*
    - Given a string s
    - Return the length of the longest substring without duplicate characters.
    Time: O(N), Space: O(n)
*/

/*
    Note:
    - sliding window keep track of what is currently considered as subarray
    - set make sure no element is repeating
*/

int lengthOfLongestSubstring(string s) {
    int left = 0;
    int right = 0;

    int maxCount = 0;
    unordered_set<char> st;
    while (left < s.size() && right < s.size()) {
        // not in pattern
        if (st.count(s[right]) == 0) {
            maxCount = max(maxCount, right - left + 1);
        } else {
            while (st.count(s[right]) > 0) {
                st.erase(s[left]);
                left++;
            }
        }
        st.insert(s[right]);
        right++;
    }

    return maxCount;
}